#### HDBSCAN

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from sklearn.metrics import adjusted_rand_score, normalized_mutual_info_score, silhouette_score, davies_bouldin_score
from collections import Counter
import hdbscan


In [2]:
data = pd.read_csv("../../../data/preprocessed/preprocessed_dataset.csv")
y = data["expression"].astype("category").cat.codes

datasets = {
    "Scaled dataset": "../../../data/reduced/X_scaled.npy",
    "PCA dataset": "../../../data/reduced/X_pca.npy"
}

In [3]:
def remap_clusters_majority(y_true, y_pred):
    """
    Mapira svaki klaster na dominantnu pravu klasu,
    radi lakseg tumacenja i eventualnog confusion matrix prikaza.
    Noise (-1) ostaje -1.
    """
    mapped = np.full_like(y_pred, fill_value=-1)

    for cluster_id in np.unique(y_pred):
        if cluster_id == -1:
            continue

        mask = (y_pred == cluster_id)
        majority_class = Counter(y_true[mask]).most_common(1)[0][0]
        mapped[mask] = majority_class

    return mapped

In [4]:
def evaluate_hdbscan(y_true, labels):
    """
    Racuna metrike samo nad tackama koje nisu oznacene kao noise.
    """
    mask = labels != -1

    n_noise = np.sum(labels == -1)
    noise_ratio = n_noise / len(labels)

    n_clusters = len(set(labels)) - (1 if -1 in labels else 0)

    if np.sum(mask) == 0 or n_clusters <= 1:
        return {
            "ari": -1,
            "nmi": -1,
            "n_clusters": n_clusters,
            "n_noise": n_noise,
            "noise_ratio": noise_ratio,
            "coverage": np.sum(mask) / len(labels)
        }

    ari = adjusted_rand_score(y_true[mask], labels[mask])
    nmi = normalized_mutual_info_score(y_true[mask], labels[mask])

    return {
        "ari": ari,
        "nmi": nmi,
        "n_clusters": n_clusters,
        "n_noise": n_noise,
        "noise_ratio": noise_ratio,
        "coverage": np.sum(mask) / len(labels)
    }


In [5]:
def plot_hdbscan_result(X, labels, dataset_name, title_suffix=""):
    plt.figure(figsize=(8, 6))
    scatter = plt.scatter(
        X[:, 0], X[:, 1],
        c=labels,
        s=8,
        cmap="tab10"
    )
    plt.title(f"HDBSCAN clustering - {dataset_name} {title_suffix}")
    plt.xlabel("Feature 1")
    plt.ylabel("Feature 2")
    plt.colorbar(scatter, label="Cluster")
    plt.tight_layout()
    plt.show()


In [ ]:
from joblib import Memory

memory = Memory("./hdbscan_cache", verbose=0)

min_cluster_sizes = [200, 500] # Predstavlja najmanju grupu koju HDBSCAN treba da razmatra kao klaster.
min_samples_values = [20, 50] # Određuje koliko se procenjuje gustina

results = []

for dataset_name, X_data in datasets.items():

    print("=" * 80)
    print(f"DATASET: {dataset_name}")
    print("=" * 80)

    X = np.load(X_data)

    # Koristimo uzorak samo za izbor parametara,
    # da ne bismo svaki put pokretali HDBSCAN nad celim datasetom.
    rng = np.random.default_rng(42)
    sample_size = min(3000, len(X))
    sample_idx = rng.choice(len(X), size=sample_size, replace=False)

    X_tune = X[sample_idx]
    y_tune = y.iloc[sample_idx].to_numpy() if hasattr(y, "iloc") else np.asarray(y)[sample_idx]

    best_selection_score = -np.inf
    best_params = None

    for min_cluster_size in min_cluster_sizes:
        for min_samples in min_samples_values:

            model = hdbscan.HDBSCAN(
                min_cluster_size=min_cluster_size,
                min_samples=min_samples,
                cluster_selection_method="eom", # Iz hijerarhije se biraju klasteri sa najvećom stabilnošću kroz različite nivoe gustine.
                memory=memory
            )

            labels = model.fit_predict(X_tune)

            non_noise_mask = labels != -1

            n_clusters = len(set(labels)) - (1 if -1 in labels else 0)
            n_noise = np.sum(labels == -1)
            noise_ratio = n_noise / len(labels)

            # Ako nema bar 2 prava klastera, Silhouette nema smisla.
            if n_clusters < 2 or np.sum(non_noise_mask) < 2:
                selection_score = -np.inf
            else:
                unique_non_noise = np.unique(labels[non_noise_mask])

                if len(unique_non_noise) < 2:
                    selection_score = -np.inf
                else:
                    silhouette = silhouette_score(
                        X_tune[non_noise_mask],
                        labels[non_noise_mask],
                        sample_size=min(500, np.sum(non_noise_mask)),
                        random_state=42
                    )

                    # Interni kriterijum:
                    # nagrađuje razdvojene klastere, kažnjava mnogo šuma.
                    selection_score = silhouette - noise_ratio

            if selection_score > best_selection_score:
                best_selection_score = selection_score
                best_params = {
                    "min_cluster_size": min_cluster_size,
                    "min_samples": min_samples
                }

    if best_params is None:
        print("Nije pronađena validna HDBSCAN konfiguracija.")
        continue

    # Finalno fitovanje na celom datasetu sa izabranim parametrima.
    final_model = hdbscan.HDBSCAN(
        min_cluster_size=best_params["min_cluster_size"],
        min_samples=best_params["min_samples"],
        cluster_selection_method="eom",
        memory=memory
    )

    final_labels = final_model.fit_predict(X)

    non_noise_mask_full = final_labels != -1
    X_valid = X[non_noise_mask_full]
    labels_valid = final_labels[non_noise_mask_full]

    n_clusters_full = len(set(final_labels)) - (1 if -1 in final_labels else 0)
    noise_points_full = np.sum(final_labels == -1)
    noise_ratio_full = noise_points_full / len(final_labels)

    if np.sum(non_noise_mask_full) >= 2 and len(np.unique(labels_valid)) >= 2:
        final_silhouette = silhouette_score(
            X_valid,
            labels_valid,
            sample_size=min(1000, len(X_valid)),
            random_state=42
        )

        final_db = davies_bouldin_score(X_valid, labels_valid)
    else:
        final_silhouette = np.nan
        final_db = np.nan

    # ARI i NMI su samo eksterna evaluacija, ne kriterijum izbora parametara.
    final_ari = adjusted_rand_score(y, final_labels)
    final_nmi = normalized_mutual_info_score(y, final_labels)

    results.append({
        "Dataset": dataset_name,
        "Algorithm": "HDBSCAN",
        "min_cluster_size": best_params["min_cluster_size"],
        "min_samples": best_params["min_samples"],
        "Clusters": n_clusters_full,
        "Noise_points": int(noise_points_full),
        "Noise_ratio": noise_ratio_full,
        "Silhouette_non_noise": final_silhouette,
        "Davies_Bouldin_non_noise": final_db,
        "ARI": final_ari,
        "NMI": final_nmi,
        "Selection_score": best_selection_score
    })

    print(f"Selected min_cluster_size: {best_params['min_cluster_size']}")
    print(f"Selected min_samples: {best_params['min_samples']}")
    print(f"Selection score: {best_selection_score:.4f}")
    print(f"ARI for selected model: {final_ari:.4f}")
    print(f"NMI for selected model: {final_nmi:.4f}")

    plt.figure(figsize=(7, 6))
    plt.scatter(X[:, 0], X[:, 1], c=final_labels, s=8, cmap="tab10")
    plt.title(f"HDBSCAN - {dataset_name}")
    plt.tight_layout()
    plt.savefig(f"../../../results/hdbscan/hdbscan_{dataset_name.replace(' ', '_').lower()}.png")
    plt.close()

results_df = pd.DataFrame(results)

results_df = results_df.round({
    "Noise_ratio": 4,
    "Silhouette_non_noise": 4,
    "Davies_Bouldin_non_noise": 4,
    "ARI": 4,
    "NMI": 4,
    "Selection_score": 4
})

results_df.to_csv("../../../results/hdbscan/hdbscan_results_summary.csv", index=False)

display(results_df.sort_values(["Dataset", "Selection_score"], ascending=[True, False]))

DATASET: Scaled dataset
Selected min_cluster_size: 200
Selected min_samples: 20
Selection score: 0.2569
ARI for selected model: 0.0268
NMI for selected model: 0.0673
DATASET: PCA dataset
Selected min_cluster_size: 200
Selected min_samples: 20
Selection score: 0.2803
ARI for selected model: 0.0221
NMI for selected model: 0.0517


,Dataset,Algorithm,min_cluster_size,min_samples,Clusters,Noise_points,Noise_ratio,Silhouette_non_noise,Davies_Bouldin_non_noise,ARI,NMI,Selection_score
1,PCA dataset,HDBSCAN,200,20,3,9142,0.3272,0.3958,0.7061,0.0221,0.0517,0.2803
0,Scaled dataset,HDBSCAN,200,20,5,9659,0.3458,0.2721,1.3551,0.0268,0.0673,0.2569


##### Primena HDBSCAN algoritma pokazuje da ovaj algoritam pronalazi mali broj gustinskih klastera, ali i značajan broj tačaka označava kao šum. Parametri su izabrani pomoću internog kriterijuma Selection_score, koji kombinuje Silhouette meru nad tačkama koje nisu označene kao šum i procenat šuma.

##### Za scaled dataset izabrani su parametri min_cluster_size = 200 i min_samples = 20, pri čemu je dobijeno 5 klastera i približno 34.6% šuma. Za PCA dataset izabrani su isti parametri, pri čemu su dobijena 3 klastera i približno 32.7% šuma.

##### PCA dataset ostvaruje bolju vrednost Silhouette_non_noise i niži Davies-Bouldin indeks, što ukazuje na bolju geometrijsku separaciju klastera u redukovanom prostoru. Međutim, niske vrednosti ARI i NMI pokazuju da dobijeni klasteri nisu dobro usklađeni sa stvarnim klasama izraza. Zbog toga se HDBSCAN može smatrati korektno primenjenim, ali ne naročito pogodnim algoritmom za izdvajanje klastera koji direktno odgovaraju gramatičkim izrazima lica.

##### HDBSCAN ostvaruje visok Silhouette, ali zbog samo 2 klastera, velikog broja šumnih tačaka i vrlo niskih ARI/NMI vrednosti, ne uspeva da pronađe strukturu koja odgovara stvarnim izrazima lica.